src/pow.rs

```diff
@@ -0,0 +1,1164 @@
+// Rust Bitcoin Library - Written by the rust-bitcoin developers.
+// SPDX-License-Identifier: CC0-1.0
+
+//! Proof-of-work related integer types.
+//!
+//! The functions here are designed to be fast.
+//!
+
+macro_rules! impl_from_and_impl {
+    ($ty:ident) => {
+        impl From<[u64; 4]> for $ty {
+            #[inline]
+            fn from(a: [u64; 4]) -> Self { $ty(Uint256(a)) }
+        }
+
+        impl From<u64> for $ty {
+            #[inline]
+            fn from(x: u64) -> Self { $ty(Uint256::from(x)) }
+        }
+
+        impl $ty {
+            /// TODO: Generate docs.
+            #[inline]
+            pub const fn from_array(a: [u64; 4]) -> $ty { $ty(Uint256(a)) }
+
+            /// TODO: Generate docs.
+            #[inline]
+            pub fn from_be_bytes(bytes: [u8; 32]) -> $ty { $ty(Uint256::from_be_bytes((bytes))) }
+
+            /// TODO: Generate docs.
+            #[inline]
+            pub fn from_be_slice(bytes: &[u8]) -> Result<$ty, ParseLengthError> {
+                let u = Uint256::from_be_slice(bytes)?;
+                Ok($ty(u))
+            }
+
+            /// TODO: Generate docs.
+            #[inline]
+            pub fn to_be_bytes(self) -> [u8; 32] { self.0.to_be_bytes() }
+        }
+    };
+}
+
+macro_rules! impl_core_traits {
+    ($ty:ident) => {
+        impl core::ops::Add<$ty> for $ty {
+            type Output = $ty;
+
+            #[inline]
+            fn add(self, other: $ty) -> $ty { $ty(self.0.wrapping_add(other.0)) }
+        }
+
+        impl core::ops::Sub<$ty> for $ty {
+            type Output = $ty;
+
+            #[inline]
+            fn sub(self, other: $ty) -> $ty { $ty(self.0.wrapping_sub(other.0)) }
+        }
+
+        impl core::ops::Mul<$ty> for $ty {
+            type Output = $ty;
+
+            #[inline]
+            fn mul(self, other: $ty) -> $ty { $ty(self.0.wrapping_mul(other.0)) }
+        }
+
+        impl core::ops::Div<$ty> for $ty {
+            type Output = $ty;
+
+            #[inline]
+            fn div(self, other: $ty) -> $ty { $ty(self.0.div(other.0)) }
+        }
+
+        impl core::ops::Rem<$ty> for $ty {
+            type Output = $ty;
+
+            #[inline]
+            fn rem(self, other: $ty) -> $ty { $ty(self.0.rem(other.0)) }
+        }
+
+        impl core::ops::BitAnd<$ty> for $ty {
```

Comment:
Bit ops kinda leak our representation, do we want to do that?

Reply:
All deleted!